src/main/kotlin/org/rust/ide/refactoring/move/common/RsMoveConflictsDetector.kt

```diff
@@ -0,0 +1,159 @@
+        fun checkVisibility(referenceElement: RsElement, target: RsVisible) {
+            if (!target.isInsideMovedElements(elementsToMove) && !target.isVisibleFrom(targetMod)) {
+                addVisibilityConflict(conflicts, referenceElement, target)
+            }
+        }
+
+        val elementsToCheck = movedElementsDeepDescendantsOfType<RsElement>(elementsToMove)
+        detectPrivateFieldOrMethodReferences(elementsToCheck, ::checkVisibility)
+    }
+
+    private fun detectPrivateFieldOrMethodReferences(
+        elementsToCheck: Sequence<RsElement>,
+        checkVisibility: (RsElement, RsVisible) -> Unit
+    ) {
+        fun checkVisibility(reference: RsReferenceElement) {
+            val target = reference.reference?.resolve() as? RsVisible ?: return
+            checkVisibility(reference, target)
+        }
+
+        loop@ for (element in elementsToCheck) {
+            when (element) {
+                is RsDotExpr -> {
+                    val fieldReference = element.fieldLookup ?: element.methodCall ?: continue@loop
+                    checkVisibility(fieldReference)
+                }
+                is RsStructLiteralField -> {
+                    val field = element.resolveToDeclaration() ?: continue@loop
+                    checkVisibility(element, field)
+                }
+                is RsPatField -> {
+                    val patBinding = element.patBinding ?: continue@loop
+                    checkVisibility(patBinding)
+                }
+                is RsPatTupleStruct -> {
+                    // it is ok to use `resolve` and not `deepResolve` here
+                    // because type aliases can't be used in destructuring tuple struct:
+                    val struct = element.path.reference?.resolve() as? RsStructItem ?: continue@loop
+                    val fields = struct.tupleFields?.tupleFieldDeclList ?: continue@loop
+                    for (field in fields) {
+                        checkVisibility(element, field)
+                    }
+                }
+                is RsPath -> {
+                    // conflicts for simple paths are handled using `pathNewAccessible`/`pathNewFallback` machinery
+                    val isInsideSimplePath = element.parentsWithSelf
+                        .takeWhile { it is RsPath }
+                        .any { isSimplePath(it as RsPath) }
+                    if (!isInsideSimplePath && !element.startsWithSelf()) {
+                        // here we handle e.g. UFCS paths: `Struct1::method1`
+                        checkVisibility(element)
+                    }
+                }
+            }
+        }
+    }
+
+    companion object {
+        val RS_ELEMENT_FOR_CHECK_INSIDE_REFERENCES_VISIBILITY: Key<RsElement> =
+            Key("RS_ELEMENT_FOR_CHECK_INSIDE_REFERENCES_VISIBILITY")
+    }
+}
+
+fun addVisibilityConflict(conflicts: MultiMap<PsiElement, String>, reference: RsElement, target: RsElement) {
```

Comment:
Yes, it will be used in [`RsMoveTraitMethodsProcessor`](https://github.com/intellij-rust/intellij-rust/blob/92b32a2758fab80c36abe8d69d97bb478c8b4a0e/src/main/kotlin/org/rust/ide/refactoring/move/common/RsMoveTraitMethodsProcessor.kt#L71) :)